src/OrchardCore/OrchardCore.Sms.Core/SmsConstants.cs

```diff
@@ -5,4 +5,6 @@ public class SmsConstants
     public const string TwilioServiceName = "Twilio";
 
     public const string ConsoleServiceName = "Console";
```

Comment:
Maybe another static class and fields so that we can use something like that `SmsProvider.Twilio`, `SmsProvider.Console`. Or maybe the sms providers should provide their names, then get the right one when resolving `ISmsProvider`, hmm but still need to register all providers in the DI and also resolve them.

Or keep the names in the options but don't register providers in the DI, only enlist types in place of resolution delegates, then activate the right one when resolving `ISmsProvider`, this by using our service provider extension.

https://github.com/OrchardCMS/OrchardCore/blob/23496769684147759d8dca7fc620381c6c93e221/src/OrchardCore/OrchardCore.Abstractions/Shell/Builders/Extensions/ServiceProviderExtensions.cs#L16-L22

Or use directly `ActivatorUtilities.CreateInstance()`.

Reply:
If you look at the way they providers are registers, we register a delegate not an instance. So the provider is not initialized until it is needed based on the DefaultProvider

Reply:
But you also have just before

        services.AddSingleton<T>();

Reply:
Also a little typo, `EnsureClientIsIsInitializedAsync()` may need to be renamed to `EnsureClientIsInitializedAsync()`.

Reply:
Yea but isn't that definitions the type in the IoC it does not resolve it. It's is resolved at run time. Isn't it?

Reply:
Nothing urgent, need to leave too, will see tomorrow.

Yes, only resolved on demand, just to not add too much descriptors to lookup, but here only a few ones are added, for info at some point we decreased scoped services in favor of singletons, but still good to decrease both.

I think that the twilio initialization need to be thread safe.